packages/react-components/src/gallery/dominantSpeaker.ts

```diff
@@ -0,0 +1,80 @@
+// Copyright (c) Microsoft Corporation.
+// Licensed under the MIT license.
+import { VideoGalleryRemoteParticipant } from '../types';
+
+/**
+ * Calculates the participants that should be renderd basis on the list of dominant
+ * speakers and currently renderd participants in a call.
+ * @param participants - Array containing all participants of a call.
+ * {@link @azure/communication-react#VideoGalleryRemoteParticipant}
+ * @param dominantSpeakers - An array containing the userId of dominant speakers
+ * in a call in the order of their dominance. 0th index is the most dominant, 1st is the second most etc
+ * @param visibleParticipants - Array containing currently rendered (visible)
+ * participants in the call. {@link @azure/communication-react#VideoGalleryRemoteParticipant}
+ * @param maxTiles - Maximum number of tiles to calculate.
+ * @returns VideoGalleryRemoteParticipant[] {@link @azure/communication-react#VideoGalleryRemoteParticipant}
+ */
+export const smartDominantSpeakerParticipants = (
+  participants: VideoGalleryRemoteParticipant[] = [],
+  dominantSpeakers: Array<string> = [],
+  visibleParticipants: VideoGalleryRemoteParticipant[] = [],
+  maxTiles = 4
+): VideoGalleryRemoteParticipant[] => {
+  if (!participants) return [];
+
+  // Don't apply any logic if total number of video streams is less than Max video streams.
+  if (participants.length <= maxTiles) return participants;
+
+  let currentParticipants = visibleParticipants.slice();
+  // Initialize `currentParticipants` if it is empty.
+  if (!currentParticipants.length) {
+    currentParticipants = participants.slice(0, maxTiles);
+  }
+
+  // Only use the Max allowed dominant speakers.
+  const dominantSpeakerIds = Array.from(new Set(dominantSpeakers).values()).slice(0, maxTiles);
+  const lastVisibleSpeakerIds = currentParticipants.map((p) => p.userId);
+  const newDominantSpeakerIds = dominantSpeakerIds.filter((id) => !lastVisibleSpeakerIds.includes(id));
+
+  // Remove participants that are no longer dominant and replace them with new dominant speakers.
+  for (let index = 0; index < maxTiles; index++) {
+    const oldUserId = lastVisibleSpeakerIds[index];
+
+    if ((oldUserId && !dominantSpeakerIds.includes(oldUserId)) || oldUserId === undefined) {
+      const replacement = newDominantSpeakerIds.shift();
+      if (!replacement) break;
+      lastVisibleSpeakerIds[index] = replacement;
+    }
+  }
+
+  // Search optimization
+  const lastVisibleSpeakersMap = {};
+  lastVisibleSpeakerIds.forEach((userId) => {
+    lastVisibleSpeakersMap[userId] = true;
+  });
+
+  const newSpeakers = participants.filter((p) => lastVisibleSpeakersMap[p.userId]);
```

Comment:
`lastVisibleSpeakersMap` is used solely here and it makes this function longer and more difficult reading than the potential perf benefits.
Could `lastVisibleSpeakersMap` be removed and instead this become:
```
const newVisibleParticipants = participants.filter((p) => lastVisibleSpeakerIds.includes(p.userId))
```

Reply:
That's how it was earlier and @prprabhu-ms  advised to add a perf optimization here.
The potential perf benefits might become significant in the near future when acs supports more participants.
I think readability is important, but we need to think of performance as well. Even if that means making an algorithm more complicated in the process.

Reply:
Hmm OK, I was thinking limits of 50 tiles and 1000 participants which should be trivial but suppose better to future proof

Reply:
Actually better to keep the algorithm agnostic of our constraints anyway